### PR TITLE
[LV] run verifyFunction once

### PR DESCRIPTION
**Pull request by artagnon**
In an effort to cut compile-times on debug builds, run verifyFunction once per function in LoopVectorize, as opposed to running it as many times as there are loops in the function. The downside of this is that, in cases of a broken function, the user won't be informed exactly which loop transformation failed, but this is perhaps an acceptable price to pay for cutting down compile-times.

Statistics from LLVM Compile Time Tracker: https://llvm-compile-time-tracker.com/compare.php?from=67c18fe6ca0d491bde18dc7f8e7c6e32e52f4dbb&to=cbfebd1b531f905efde60c913c9dd35b6b161019&stat=instructions:u

### CODE REVIEW

**Review by david-arm — APPROVED** [review]
LGTM. I think this change makes sense. Although it doesn't help release builds, I imagine it does help reduce the amount of test time in the buildbots.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 1]
```diff
@@ -10359,6 +10357,8 @@ LoopVectorizeResult LoopVectorizePass::runImpl(
     }
   }
 
+  assert(!Changed || !verifyFunction(F, &dbgs()));
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:10360]: `Changed` may also be set if the loop only has been modified by `simplifyLoop` or `formLCSSARecursively`, so we will verify in cases where we didn't before?

**Review by fhahn — CHANGES_REQUESTED** [review]
Do you have any data motivating the change? As @david-arm this won't help with the compile-time-tracker, as I think it only uses release builds without assertions.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 1 above)
Comment (david-arm) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:10360]: Ah sorry @fhahn, I hadn't appreciated that we might end up verifying even when not vectorising. That's potentially a compile-time regression in some cases then.

**Comment by artagnon** [comment]
Oh no, if the data doesn't show up in the LLVM Compile Time Tracker, then I'm afraid I don't have any reliable data: I'm on a MacBook, and there doesn't seem to be any reliable way to get performance benchmarking data without perf-tools or callgrind. The patch is very much a blind experiment, as I noticed wasteful verification in Instruments.app when building CTMark. I posted the patch because the downside is quite minimal.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 1 above)
Comment (artagnon) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Sorry about this; fixed now.

**Comment by artagnon** [comment]
I managed to do some statistical benchmarking of this patch, and the results are good:

| Command | Mean [s] | Min [s] | Max [s] |
|:---|---:|---:|---:|
| main | 8.471 ± 0.012 | 8.457 | 8.495 |
| lv-verify-once | 8.462 ± 0.015 | 8.446 | 8.485 |

**Comment by nikic** [comment]
Compile-time tracker results (for an assertion enabled build): https://llvm-compile-time-tracker.com/compare.php?from=67c18fe6ca0d491bde18dc7f8e7c6e32e52f4dbb&to=cbfebd1b531f905efde60c913c9dd35b6b161019&stat=instructions:u

And as a related curiosity, it seems like assertion-enabled builds have about ~50% mean overhead relative to normal release builds: https://llvm-compile-time-tracker.com/compare.php?from=e2d17a053edff68c9761aceb8ff8303e3e37172c&to=67c18fe6ca0d491bde18dc7f8e7c6e32e52f4dbb&stat=instructions:u

File: llvm/CMakeLists.txt [hunk 2]
```diff
@@ -624,7 +624,7 @@ option(LLVM_ENABLE_DUMP "Enable dump functions even when assertions are disabled
 option(LLVM_UNREACHABLE_OPTIMIZE "Optimize llvm_unreachable() as undefined behavior (default), guaranteed trap when OFF" ON)
 
 if( NOT uppercase_CMAKE_BUILD_TYPE STREQUAL "DEBUG" )
-  option(LLVM_ENABLE_ASSERTIONS "Enable assertions" OFF)
+  option(LLVM_ENABLE_ASSERTIONS "Enable assertions" ON)
```
Comment (david-arm) [inline llvm/CMakeLists.txt:627]: Depending upon what @fhahn thinks this patch looks reasonable to me - reducing compile time even for an assertion build should help our buildbots. The results @nikic posted for the compile-time tracker looked good. I guess you'd need to revert this change first though.

**Comment by artagnon** [comment]
After some more thought, I realized that there is quite a steep cost to this change that I didn't think about when authoring the patch: it will slow down the development of LoopVectorize, and contributors will be greatly inconvenienced by having to go through the entire function's output to find the breakage. It primarily benefits the buildbot, but I don't think it justifies this cost.
